hera_cal/utils.py

```diff
@@ -1033,6 +1033,87 @@ def gain_relative_difference(old_gains, new_gains, flags, denom=None):
     return relative_diff, avg_relative_diff
 
 
+def red_average(hd, reds=None, bl_tol=1.0, wgt_by_int=False, inplace=False):
+    """
+    Redundantly average visibilities in a HERAData or UVData object.
+
+    Args:
+        hd : HERAData or UVData object
+            A UVData subclass object to redundantly average
+        reds : list, optional
+            Nested lists of antpair tuples to redundantly average.
+            E.g. [ [(1, 2), (2, 3)], [(1, 3), (2, 4)], ...]
+            If None, will calculate these from the metadata
+        bl_tol : float
+            Baseline redundancy tolerance in meters. Only used if reds is None.
+        wgt_by_int : bool
+            Weight average by total integration time (nsamples * integration_time * ~flags)
+            Otherwise weighting is just uniform (1 * ~flags)
+        inplace : bool
+            Perform average and downselect inplace, otherwise returns a deepcopy.
+            The first baseline in each reds sublist is kept.
+
+    Returns:
+        HERAData : if inplace is False
+    """
+    from hera_cal import redcal
+
+    # deepcopy
+    if not inplace:
+        hd = copy.deepcopy(hd)
+
+    # get metadata
+    pols = [polnum2str(pol) for pol in hd.polarization_array]
+
+    # get redundant groups
+    if reds is None:
+        antpos, ants = hd.get_ENU_antpos()
+        antposd = dict(zip(ants, antpos))
+        reds = redcal.get_pos_reds(antposd, bl_error_tol=bl_tol)
+
+    # eliminate baselines not in data
+    antpairs = hd.get_antpairs()
+    reds = [[bl for bl in blg if bl in antpairs] for blg in reds]
+    reds = [blg for blg in reds if len(blg) > 0]
+
+    # iterate over redundant groups and polarizations
+    for pol in pols:
+        for blg in reds:
+            # get data and weight arrays for this pol-blgroup
+            d = np.asarray([hd.get_data(bl + (pol,)) for bl in blg])
+            f = np.asarray([(~hd.get_flags(bl + (pol,))).astype(np.float) for bl in blg])
+            n = np.asarray([hd.get_nsamples(bl + (pol,)) for bl in blg])
+            tint = np.asarray([hd.integration_time[hd.antpair2ind(bl + (pol,))] for bl in blg])[:, :, None]
+            if wgt_by_int:
```

Comment:
I think that's true, however, I don't want to necessitate that information being present in order to proceed. Really what we would want is the Tsys for each baseline, regardless of how its attained. However, I think it then gets a little muddy on what Tsys to assume for the averaged data (you would need to take the Tsys weighted Tsys of each individual bl, I think). I don't anticipate actually doing a redundant average weighted by Tsys, so my inclination is to merge this functionality in now, and if we see that we want to do that incorporate that functionality when needed. is that okay?

Reply:
I'm OK with that, but let's make it an issue.

Another alternative is to allow the users to provide linear weights manually as a DataContainer (instead of flags/nsamples). And if wgts are provided, then flags/nsamples are ignored. I think that's my preference, since it brings it more in line with how other functions in `hera_cal.utils` work, like `chisq`.

Reply:
I see, that gets a little complicated though b/c there isn't a simple way to do that for `UVData` inputs. I think if its okay I'll default to keeping it as is but will make an issue in case we feel that this is something we want to do, seeing as it is unlikely we will do this in the near term.